File: web/src/engine/ball.js
// Ball physics

import { PITCH } from './pitch.js';

export class Ball {
  constructor() {
    this.radius = 3;
    this.reset();
  }

  reset() {
    this.x = PITCH.x + PITCH.w / 2;
    this.y = PITCH.y + PITCH.h / 2;
    this.vx = 0;
    this.vy = 0;
    this.wasKicked = false;
  }

  update(dt) {
    const friction = 0.98;
    this.vx *= friction;
    this.vy *= friction;

    if (Math.abs(this.vx) < 0.01) this.vx = 0;
    if (Math.abs(this.vy) < 0.01) this.vy = 0;

    this.x += this.vx * dt * 60;
    this.y += this.vy * dt * 60;

    // Bounce off top/bottom
    if (this.y - this.radius < PITCH.y) {
      this.y = PITCH.y + this.radius;
      this.vy = Math.abs(this.vy) * 0.7;
    }
    if (this.y + this.radius > PITCH.y + PITCH.h) {
      this.y = PITCH.y + PITCH.h - this.radius;
      this.vy = -Math.abs(this.vy) * 0.7;
    }

    // Bounce off left/right (except goal openings)
    const inGoalY =
      this.y > PITCH.y + (PITCH.h - 40) / 2 &&
      this.y < PITCH.y + (PITCH.h + 40) / 2;

    if (this.x - this.radius < PITCH.x && !inGoalY) {
      this.x = PITCH.x + this.radius;
      this.vx = Math.abs(this.vx) * 0.7;
    }
    if (this.x + this.radius > PITCH.x + PITCH.w && !inGoalY) {
      this.x = PITCH.x + PITCH.w - this.radius;
      this.vx = -Math.abs(this.vx) * 0.7;
    }
  }

  kick(angle, power) {
    this.vx += Math.cos(angle) * power;
    this.vy += Math.sin(angle) * power;
    this.wasKicked = true;
  }

  draw(ctx) {
    // Shadow
    ctx.fillStyle = 'rgba(0,0,0,0.2)';
    ctx.fillRect(
      Math.floor(this.x - this.radius),
      Math.floor(this.y - this.radius + 2),
      this.radius * 2,
      this.radius * 2
    );

    // White ball
    ctx.fillStyle = '#ffffff';
    ctx.beginPath();
    ctx.arc(Math.floor(this.x), Math.floor(this.y), this.radius, 0, Math.PI * 2);
    ctx.fill();

    // Pattern
    ctx.fillStyle = '#333333';
    ctx.fillRect(Math.floor(this.x) - 1, Math.floor(this.y) - 1, 2, 2);
  }
}

File: server/internal/game/ball.go
package game

import "math"

type BallState struct {
	X      float64 `json:"x"`
	Y      float64 `json:"y"`
	VX     float64 `json:"vx"`
	VY     float64 `json:"vy"`
	Radius float64 `json:"radius"`
}

func NewBall() BallState {
	return BallState{X: PitchX + PitchW/2, Y: PitchY + PitchH/2, Radius: 3}
}

func (b *BallState) Reset() {
	b.X = PitchX + PitchW/2
	b.Y = PitchY + PitchH/2
	b.VX = 0
	b.VY = 0
	b.Radius = 3
}

func (b *BallState) Kick(angle, power float64) {
	b.VX += math.Cos(angle) * power
	b.VY += math.Sin(angle) * power
}

func (b *BallState) Update(dt float64) {
	b.VX *= 0.98
	b.VY *= 0.98
	if math.Abs(b.VX) < 0.01 {
		b.VX = 0
	}
	if math.Abs(b.VY) < 0.01 {
		b.VY = 0
	}
	b.X += b.VX * dt * 60
	b.Y += b.VY * dt * 60

	if b.Y-b.Radius < PitchY {
		b.Y = PitchY + b.Radius
		b.VY = math.Abs(b.VY) * 0.7
	}
	if b.Y+b.Radius > PitchY+PitchH {
		b.Y = PitchY + PitchH - b.Radius
		b.VY = -math.Abs(b.VY) * 0.7
	}
	inGoalY := b.Y > goalTop && b.Y < goalBot
	if b.X-b.Radius < PitchX && !inGoalY {
		b.X = PitchX + b.Radius
		b.VX = math.Abs(b.VX) * 0.7
	}
	if b.X+b.Radius > PitchX+PitchW && !inGoalY {
		b.X = PitchX + PitchW - b.Radius
		b.VX = -math.Abs(b.VX) * 0.7
	}
}
